### DISCUSSION TITLE
How to reload environment variables during runtime?

### DISCUSSION THREAD

**Discussion by marceloschmidt** [vercel/vercel#4920] [Help] [answered]
I have a use case where I might run into concurrency issues. I've fixed the concurrency problem, but here's what I'm dealing with now:

My project uses an API token to connect to a third-party system.
If I make one call to the project, it will check env variables for the access token, check if it's valid, and if not, generate a new access token, and store it in the environment variables. Next call to the project will use the newly updated variable, so far so good.

Problem is when the access token is invalid and I make two simultaneous calls. What would naturally happen in this case is that both calls would produce new access tokens, and store them one at a time in the environment variables. This would not be a problem if I could generate as many access tokens as needed, which is not the case. 

So I figured I'd implement a locking mechanism so that only one of the calls would generate a new access token. The second call would hit the lock and wait until the lock is released, then it would fetch the access token from the environment variable that the first execution set. Problem is that environment variables are set once the process is started... I know I can use the APIs to fetch the environment variables, and I know I can use the `?decrypt` parameter to decrypt a variable, but this only works with variables set in the `development` region.

Ideas on how to fetch the environment variable during runtime? I will just use the development region if there's no better way of doing it...

**Comment by styfle** [comment, marked as answer]
This sounds like a job for session state. 

Try using Redis to read/write state in your serverless functions. 

You can fetch the token from Redis and assign to “process.env”.

**Reply by marceloschmidt** [reply]
Thank you. Yes, I think that works :)
